mkosi/__init__.py

```diff
@@ -1727,6 +1730,42 @@ def build_uki(
             ]  # fmt: skip
         else:
             arguments += ["--pcr-private-key", context.config.sign_expected_pcr_key]
+    elif ArtifactOutput.pcrs in context.config.split_artifacts:
+        assert context.config.sign_expected_pcr_certificate
+
+        arguments += [
+            "--json=short",
+            "--policy-digest",
+            "--pcr-banks", "sha256",
+        ]  # fmt: skip
+
+        # If we are given an x509 certificate we need to extract the public key to use it
+        if context.config.sign_expected_pcr_certificate.read_text().startswith(
+            "-----BEGIN CERTIFICATE-----"
+        ):
+            keyutil = context.config.find_binary("systemd-keyutil", "/usr/lib/systemd/systemd-keyutil")
+            if not keyutil:
+                die("Could not find systemd-keyutil")
+            public_key = context.workspace / "pcr-public-key.pem"
+            keyutil_arguments: list[PathString] = [
+                keyutil,
+                "public",
+                "--certificate", context.config.sign_expected_pcr_certificate,
+            ]  # fmt: skip
+            if context.config.sign_expected_pcr_certificate_source.type == CertificateSourceType.provider:
+                keyutil_arguments += [
+                    "--certificate-source", f"provider:{context.config.sign_expected_pcr_certificate_source.source}",  # noqa: E501
+                ]  # fmt: skip
+            run(keyutil_arguments, stdout=public_key.open("wb"))
+            arguments += ["--pcr-public-key", workdir(public_key)]
+            options += ["--ro-bind", public_key, workdir(public_key)]
+        else:
+            arguments += [
+                "--pcr-public-key", workdir(context.config.sign_expected_pcr_certificate),
+            ]  # fmt: skip
+            options += [
+                "--ro-bind", context.config.sign_expected_pcr_certificate, workdir(context.config.sign_expected_pcr_certificate),  # noqa: E501
+            ]  # fmt: skip
```

Comment:
I don't see the point of having to do this inside of `mkosi`, `ukify` calls `systemd-keyutil` itself to extract the public key, why do it again here

Reply:
Sure but systemd-measure is invoked by ukify? What I'm saying is that ukify already knows how to invoke systemd-keyutil to get the public key. If it doesn't do that yet for --policy-digest, we should fix ukify to do it for that codepath as well.

Reply:
ok, stampity then https://github.com/systemd/systemd/pull/36321